README
-*- text -*-

************************************
*** madlld -- MAD low-level      ***  v 1.1, 2004-02-22
*** demonstration/decoder        ***  (c) 2001--2004 Bertrand Petit
************************************
    Release notes


This program is a "how to" about libmad's low-level API. The intended
use of this program is to serve as tutorial or demonstration for
programers wiling to learn how the curently undocumented library API
can be used. As a side effect the program can be used to decode MPEG
audio streams to raw PCM samples.

Madlld was written on an unix system, dependencies on the posix
interface are handled by providing an alternate behaviour on non-unix
systems. The hosting operating system should nontheless provide
standard streams redirection. The program structure is very simple: it
reads an audio MPEG stream from stdin and writes raw big-endian stereo
16 bits signed samples on stdout. The source code is heavily
commented, you are advised to read it bottom-up as that's were this
package purpose is.


==============================================================================
Release status
==============================================================================

This second release was tested against MAD versions 0.14.0b, 0.14.1b,
0.14.2b, 0.15.0b, and 0.15.1b. There is no known bug.


==============================================================================
Installation
==============================================================================

This program is distributed in source code only, you shall build it
first. Follow the instructions bellow to build and install malld on
unix systems. 

1. Be sure that the MAD suite is installed on your system. As of the
   14th of January 2003, the MAD suite version 0.15.0b can be freely
   obtained from <URL:http://www.underbit.com/products/mad/>. The
   suite is composed of three archives: libid3tag, libmad, and
   madplay. Madlld use only libmad. 

2. Configure the build process by editing the Makefile. Inlined
   comments will guide you there.

3. Run "make". If you are using gcc, you can safely ignore warnings
   about the use of braced-groups and the "asm" keyword. Theses are
   part of macros defined by the mad.h header.

4. Test the built binaries by placing an audio MPEG stream in the
   build directory as a file named "test.mp3". If sox(1) is intalled
   on your system them type "make test". Else type "make test.cdr"
   and use any tools of your choice to check the PCM samples that
   were store in the file "test.cdr".

5. If step 3 suceeded then you're ready to install the program with
   "make install".

6. If your brand new installation required some changes to the program,
   please send unified diffs to the author (see the Contact section
   bellow).

7. If your brand new installation was done on an operating system not
   listed in this document or against an also unlisted MAD version,
   please report your system details/versions to the author (see the
   Contact section bellow).

8. Optionally you can choose to build the postscript documentation.
   You'll need the a2ps command (version 4.10 or up), and the psutils
   package. You can obtain the two of then at theses locations:

       a2ps
         http://www.inf.enst.fr/~demaille/a2ps/

       psutils
         http://gershwin.ens.fr/vdaniel/Doc-Locale/Outils-Gnu-Linux/PsUtils/

   You can also use the preformated documents available at the madlld
   distribution site: http://www.bsd-dk.dk/~elrond/audio/madlld/#docs


Users of non-unix systems wishing to compile this program are on their
own. They must arrange to compile the madlld.c and bstdfile.c modules,
then link the resulting object files against the MAD library.

The compilation should yield no trouble as most of the program was
written in an almost pure ANSI C fashion. The only part that diverges
from the specs is the function dealing with command-line parsing. On
unix systems we use the getopt() function; a crude replacement is
provided if your system does not provide it. The use the getopt part
or the alternative is governed by the detection at pre-processing time
of several well known pre-defined macros present on unix systems. If
your preprocessor does not define theses macros and you have the
getopt() function you can still use it if you arrange to pre-define to
any value the HAVE_GETOPT macro.

When the alternative command-line parser is used, the usage rules
listed in the manual page are not applicable. For more information
just invoque the madlld filter with any number bogus non-numeric
arguments.


==============================================================================
Contact
==============================================================================

Questions, bug reports, patches, etc. are welcome. Please mail your
queries to < madlld at phoe dot frmug dot org >. If you value your
privacy you can send OpenPGP crypted mails, the author's key is
available on public keyrings under the 0606CEF7 ID, it has the
following fingerprint:

    79C9 10C8 9C3A 3F35 482F  06D6 150B A8BB 0606 CEF7

Note that things such as HTML e-mails, word documents, etc. will be
blindly discarded.

Questions related to the MAD suite should be directed to the mad-dev
mailing list. You can subscribe to the reflector through the HTTP
interface located at <URL:http://www.mars.org/bin/mailman/listinfo/mad-dev>.


==============================================================================
Fine print
==============================================================================

This program is (c) 2001--2004 by Bertrand Petit. It is distributed
under the terms of the license similar to the Berkeley license
reproduced bellow.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.

3. Neither the name of the author nor the names of its contributors
   may be used to endorse or promote products derived from this software
   without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS
BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


------------------------------------------------------------------------------
End of file README
------------------------------------------------------------------------------


CHANGELOG
-*- text -*-

------------------------------------------------------------------------------
madlld v1.1 (2004-02-22)
------------------------------------------------------------------------------

* Added an explanation about the need for appending MAD_BUFFER_GUARD
  bytes to the input buffer at the end of streams.

* Added simple examples on the alteration of subband domain samples
  for audio filtering.

* Added the demonstration of the use of the MAD_EMPHASIS_RESERVED
  enumeration value, new in libmad v0.15.0b.

* The fixed num to short conversion: uses signed integer numbers as it
  should had been. The numbers were signed but the C data types were
  unsigned. Added over range samples clipping.

* Automated the production of the postscript documentation.

* Fixed some typos.

------------------------------------------------------------------------------
madlld v1.0p1 (2002-01-08)
------------------------------------------------------------------------------

* Just reworked some comments to clarify the MAD_ERROR_BUFLEN error
  condition and the partial frames at the end of the input buffer.

* Added a warning about the unsatifying quality of the
  MadFixedToUshort() function.

* Fixed some typos.

------------------------------------------------------------------------------
madlld v1.0 (2001-11-13)
------------------------------------------------------------------------------

* Initial release.

EOF

bstdfile.h
/* HTAB = 4 */
/****************************************************************************
 * bstdfile.h -- This module implements a buffered interface for the		*
 * fread(2)	standard function that can signal an eof of file condition		*
 * synchronously with the transmission of the last bytes of a file.			*
 *--------------------------------------------------------------------------*
 * (c) 2004 Bertrand Petit													*
 *																			*
 * Redistribution and use in source and binary forms, with or without		*
 * modification, are permitted provided that the following conditions		*
 * are met:																	*
 *																			*
 * 1. Redistributions of source code must retain the above copyright		*
 *    notice, this list of conditions and the following disclaimer.			*
 *																			*
 * 2. Redistributions in binary form must reproduce the above				*
 *    copyright notice, this list of conditions and the following			*
 *    disclaimer in the documentation and/or other materials provided		*
 *    with the distribution.												*
 * 																			*
 * 3. Neither the name of the author nor the names of its contributors		*
 *    may be used to endorse or promote products derived from this			*
 *    software without specific prior written permission.					*
 * 																			*
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS''		*
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED		*
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A			*
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR		*
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,				*
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT			*
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF			*
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND		*
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,		*
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT		*
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF		*
 * SUCH DAMAGE.																*
 *																			*
 ****************************************************************************/

/*
 * $Name: v1_1 $
 * $Date: 2004/02/13 17:58:22 $
 * $Revision: 1.1 $
 */

#ifndef BSTDFILE_H
#define BSTDFILE_H

/****************************************************************************
 * Datatypes definitions													*
 ****************************************************************************/
typedef struct bstdfile bstdfile_t;

/****************************************************************************
 * Prototypes																*
 ****************************************************************************/
extern bstdfile_t	*NewBstdFile(FILE *fp);
extern int			BstdFileDestroy(bstdfile_t *BstdFile);
extern int			BstdFileEofP(const bstdfile_t *BstdFile);
extern int			BstdFileErrorP(const bstdfile_t *BstdFile);
extern size_t		BstdRead(void *UserBuffer, size_t ElementSize, size_t ElementsCount, bstdfile_t *BstdFile);

#endif /* BSTDFILE_H */

/*
 * Local Variables:
 * tab-width: 4
 * End:
 */

/****************************************************************************
 * End of file bstdflle.h													*
 ****************************************************************************/


bstdfile.c
/* HTAB = 4 */
/****************************************************************************
 * bstdfile.h -- This module implements a buffered interface for the		*
 * fread(2)	standard function that can signal an eof of file condition		*
 * synchronously with the transmission of the last bytes of a file.			*
 *--------------------------------------------------------------------------*
 * (c) 2004 Bertrand Petit													*
 *																			*
 * Redistribution and use in source and binary forms, with or without		*
 * modification, are permitted provided that the following conditions		*
 * are met:																	*
 *																			*
 * 1. Redistributions of source code must retain the above copyright		*
 *    notice, this list of conditions and the following disclaimer.			*
 *																			*
 * 2. Redistributions in binary form must reproduce the above				*
 *    copyright notice, this list of conditions and the following			*
 *    disclaimer in the documentation and/or other materials provided		*
 *    with the distribution.												*
 * 																			*
 * 3. Neither the name of the author nor the names of its contributors		*
 *    may be used to endorse or promote products derived from this			*
 *    software without specific prior written permission.					*
 * 																			*
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS''		*
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED		*
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A			*
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR		*
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,				*
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT			*
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF			*
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND		*
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,		*
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT		*
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF		*
 * SUCH DAMAGE.																*
 *																			*
 ****************************************************************************/

/*
 * $Name: v1_1 $
 * $Date: 2004/02/13 17:58:22 $
 * $Revision: 1.1 $
 */

/****************************************************************************
 * Includes.																*
 ****************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/types.h>
#include "bstdfile.h"

/****************************************************************************
 * Preprocessor definitions													*
 ****************************************************************************/
#define BFILE_BUFSIZE	(8192U)

/* We use some errno symbols that may be undefined. We supply such
 * definitions, all with the 1 value, only to acheive compilation: the
 * semantics of the signaled errors may thus be broken. It is the
 * reponssibility of this module user to map the following symbols to
 * the symbols provided by the target system.
 */
#if (!defined(unix) && !defined (__unix__) && !defined(__unix))
/* Memory exhausted */
# ifndef ENOMEM
#  define ENOMEM	(1)
# endif

/* Bad file descriptor */
# ifndef EBADF
#  define EBADF		(1)
# endif

/* Invalid address. */
# ifndef EFAULT
#  define EFAULT	(1)
# endif

/* Invalid argument. */
# ifndef EINVAL
#  define EINVAL	(1)
# endif
#endif

/****************************************************************************
 * Datatypes definitions													*
 ****************************************************************************/
struct bstdfile
{
	/* buffer is an internal buffer used by BstdRead(). live points
	 * inside that buffer to the data tha was not yet transmited to
	 * the user. live_size is the number of bytes available for direct
	 * conspumtion to feed the user buffer from the live pointer.
	 */
	char	buffer[BFILE_BUFSIZE],
			*live;
	size_t	live_size;

	/* fp is the file openned for reading associated with that bfile. */
	FILE	*fp;

	/* Error managment: eof is non-zero when an end of file condition
	 * was detected. error is zero when no error was detected (error
	 * is zero when eof is 1), it stores the errno of the detected
	 * error.
	 */
	int		eof,
			error;
};

/****************************************************************************
 * Creates a new bstdfile from an already openned file.						*
 ****************************************************************************/
bstdfile_t *NewBstdFile(FILE *fp)
{
	bstdfile_t	*BstdFile;

	/* Allocate the bstdfile structure. */
	BstdFile=(bstdfile_t *)malloc(sizeof(bstdfile_t));
	if(BstdFile==NULL)
	{
		errno=ENOMEM;
		return(NULL);
	}

	/* Initialize the structure to safe defaults. */
	BstdFile->live=BstdFile->buffer;
	BstdFile->live_size=0;
	BstdFile->eof=0;
	BstdFile->error=0;
	BstdFile->fp=fp;

	/* Return the new bfile. */
	return(BstdFile);
}

/****************************************************************************
 * Destroys a previously allocated BstdFile.								*
 ****************************************************************************/
int BstdFileDestroy(bstdfile_t *BstdFile)
{
	if(BstdFile==NULL)
	{
		errno=EBADF;
		return(1);
	}
	free(BstdFile);
	return(0);
}

/****************************************************************************
 * This predicate returns a non nul value when there is an end of file		*
 * condition on the BstdFile argument.										*
 ****************************************************************************/
int BstdFileEofP(const bstdfile_t *BstdFile)
{
	return(BstdFile->eof);
}

/****************************************************************************
 * This predicate returns a non nul value when there is an error condition	*
 * on the BstdFile argument.												*
 ****************************************************************************/
int BstdFileErrorP(const bstdfile_t *BstdFile)
{
	return(BstdFile->error);
}

/****************************************************************************
 * This works as read(2) but operates on a bfile instead of a file			*
 * descriptor.																*
 ****************************************************************************/
size_t BstdRead(void *UserBuffer, size_t ElementSize, size_t ElementsCount, bstdfile_t *BstdFile)
{
	size_t	RequestSize=ElementSize*ElementsCount,
			FeededSize=0,
			ReadSize,
			ObtainedSize;
	int		OldErrno=errno;

	/* Check the validity of the arguments. */
	if(BstdFile==NULL)
	{
		errno=EBADF;
		return((size_t)0);
	}
	if(UserBuffer==NULL)
	{
		errno=EFAULT;
		return((size_t)0);
	}
	if(RequestSize<1)
	{
		errno=EINVAL;
		return((size_t)0);
	}

	/* Return immediately if an exceptionnal situation exists. */
	if(BstdFile->eof)
		return((size_t)0);
	if(BstdFile->error)
	{
		errno=BstdFile->error;
		return((size_t)0);
	}

	/* The easy case. */
	if(RequestSize==0U)
		return((size_t)0);

	/* First feed the target buffer from the BstdFile buffer if it has
	 * some meat to be feeded on.
	 */
	if(BstdFile->live_size>0)
	{
		/* If there is more data in the buffer than requested by the
		 * user then we feed him directly from our buffer without a
		 * read operation.
		 */
		if(BstdFile->live_size>RequestSize)
		{
			memcpy(UserBuffer,BstdFile->live,RequestSize);
			BstdFile->live+=RequestSize;
			BstdFile->live_size-=RequestSize;
			return(RequestSize);
		}
		/* Else we drain our buffer. */
		else
		{
			memcpy(UserBuffer,BstdFile->live,BstdFile->live_size);
			UserBuffer=(char *)UserBuffer+BstdFile->live_size;
			FeededSize=BstdFile->live_size;
			BstdFile->live=BstdFile->buffer;
			BstdFile->live_size=0;
		}
	}

	/* If the user request was not yet fulfilled we then read from the
     * file the remaining data requested by the user.
	 */
	if(FeededSize<RequestSize)
	{
		ReadSize=RequestSize-FeededSize;
		ObtainedSize=fread(UserBuffer,1,ReadSize,BstdFile->fp);
		FeededSize+=ObtainedSize;

		/* If an error occurs we return the amount of data that was
		 * feeded from the buffer and store the error condition for a
		 * later call. If our buffer was empty and we thus have
		 * transferred no data to the user buffer then we directly
		 * return the error.
		 */
		if(ObtainedSize==0U)
		{
			if(feof(BstdFile->fp))
				BstdFile->eof=1;
			else
			{
				BstdFile->error=errno;
				errno=OldErrno;
			}
			if(FeededSize!=0)
				return(FeededSize);
			else
				return(0U);
		}
	}

	/* Fill again our buffer. In case of error, or end of file, that
	 * error is recorded but we still report the amount of data that
	 * was feeded to the user buffer.
	 */
	ObtainedSize=fread(BstdFile->buffer,1,BFILE_BUFSIZE,BstdFile->fp);
	if(ObtainedSize==0)
	{
		if(feof(BstdFile->fp))
			BstdFile->eof=1;
		else
		{
			BstdFile->error=errno;
			errno=OldErrno;
		}
	}
	else
	{
		BstdFile->live=BstdFile->buffer;
		BstdFile->live_size=ObtainedSize;
	}

	/* Eventually return the number ob bytes feeded to the user
     * buffer.
	 */
	return(FeededSize);
}

/*
 * Local Variables:
 * tab-width: 4
 * End:
 */

/****************************************************************************
 * End of file bflle.c														*
 ****************************************************************************/
